File: app/models/message.rb
# frozen_string_literal: true

# == Schema Information
#
# Table name: messages
#
#  id         :bigint           not null, primary key
#  account_id :bigint
#  chat_id    :bigint           not null
#  role       :integer          default("user"), not null
#  content    :text             not null
#  created_at :datetime         not null
#  updated_at :datetime         not null
#
# Message belonging to a chat
class Message < ApplicationRecord
  belongs_to :chat, touch: :latest_message_at
  belongs_to :account, touch: :latest_message_at, optional: true

  validates :content, presence: true
  validates :account_id, presence: true, if: :user?

  enum :role, { user: 0, assistant: 1, system: 2 }
end

File: channel/lib/channel/chats/message.ex
defmodule Channel.Chats.Message do
  @moduledoc """
  Chat Messages
  """

  use Ecto.Schema
  import Ecto.Changeset
  alias Channel.{Auth.Account, Chats.Chat}

  schema "messages" do
    field :role, :integer, default: 0
    field :content, :string

    belongs_to :account, Account
    belongs_to :chat, Chat

    timestamps(type: :utc_datetime, inserted_at: :created_at)
  end

  def changeset(message, attrs) do
    message_cols = [:chat_id, :account_id, :content]

    message
    |> cast(attrs, message_cols)
    |> validate_required(message_cols)
  end
end
